pennylane/io/to_openqasm.py

```diff
@@ -0,0 +1,327 @@
+    if op.reset:
+        raise NotImplementedError(f"Unable to translate mid circuit measurements with reset {op}.")
+    if op.postselect:
+        raise NotImplementedError(
+            f"Unable to translate mid circuit measurement with postselection {op}"
+        )
+    wire = f"q[{wires.index(op.wires[0])}]"
+    mcm_ind = len(bit_map)
+    bit_map[op] = mcm_ind
+    return f"measure {wire} -> mcms[{mcm_ind}];"
+
+
+def _conditional_str(op: Conditional, wires: Wires, bit_map: dict, precision: None | int) -> str:
+    if op.meas_val.has_processing:
+        raise NotImplementedError(
+            "to_openqasm does not support translating Conditionals with measurement postprocessing."
+        )
+    mcm_name = f"mcms[{bit_map[op.meas_val.measurements[0]]}]"
+    return f"if({mcm_name}==1) {_operator_str(op.base, wires, bit_map, precision)}"
+
+
+# pylint: disable=unused_argument
+def _operator_str(op: Operator, wires: Wires, bit_map: dict, precision: None | int) -> str:
+    try:
+        gate = OPENQASM_GATES[op.name]
+    except KeyError as e:
+        raise ValueError(f"Operation {op.name} not supported by the QASM serializer") from e
+
+    wire_labels = ",".join([f"q[{wires.index(w)}]" for w in op.wires.tolist()])
+    params = ""
+
+    if op.num_params > 0:
+        # If the operation takes parameters, construct a string
+        # with parameter values.
+        if precision is not None:
+            params = "(" + ",".join([f"{p:.{precision}}" for p in op.parameters]) + ")"
+        else:
+            # use default precision
+            params = "(" + ",".join([str(p) for p in op.parameters]) + ")"
+
+    return f"{gate}{params} {wire_labels};"
+
+
+def _tape_openqasm(
+    tape: QuantumScript, wires: Wires, rotations: bool, measure_all: bool, precision: None | int
+):
+    """Helper function to serialize a tape as an OpenQASM 2.0 program."""
+    wires = wires or tape.wires
+
+    # add the QASM headers
+    lines = ["OPENQASM 2.0;", 'include "qelib1.inc";']
+
+    if tape.num_wires == 0:
+        # empty circuit
+        return "\n".join(lines) + "\n"
+
+    # create the quantum and classical registers
+    lines.append(f"qreg q[{len(wires)}];")
+    lines.append(f"creg c[{len(wires)}];")
+
+    num_mcms = sum(isinstance(o, MidMeasureMP) for o in tape.operations)
+    if num_mcms:
+        lines.append(f"creg mcms[{num_mcms}]")
+    bit_map = {}
+
+    # get the user applied circuit operations without interface information
+    [transformed_tape], _ = convert_to_numpy_parameters(tape)
+    operations = transformed_tape.operations
+
+    if rotations:
+        # if requested, append diagonalizing gates corresponding
+        # to circuit observables
+        operations += tape.diagonalizing_gates
+
+    just_ops = QuantumScript(operations)
+
+    def stopping_condition(op):
+        return op.name in OPENQASM_GATES or isinstance(op, (MidMeasureMP, Conditional))
+
+    [new_tape], _ = decompose(
```

Comment:
Now uses decompose instead of `tape.expand` :) Hooray to have one less place that uses tape expand.